### PR TITLE
added keystore support for custom functions auth

### LINKED ISSUE #3868: Add credential helpers support for composition functions
### What problem are you facing?
Using ECR for my customm Composition Functions. crossplane-xfn supports only static credentials, which are not applicable for ECR registries

### How could Crossplane help solve your problem?
Add go-containerregistry multi-keychain support to support default helpers for ECR, ACR, GCR, Docker Hub, etc

### PR DESCRIPTION
**Pull request by AndrewChubatiuk**
### Description of your changes

**crossplane-xfn** doesn't support pulling images from:
- registries such as ECR, ACR, which are using temporary credentials
- service account credentials
- kube secrets

Fixes #3868 #3717

I have:

- [x] Read and followed Crossplane's [contribution process].
- [x] Run `make reviewable` to ensure this PR is ready for review.
- [x] Added `backport release-x.y` labels to auto-backport this PR if necessary.

### How has this code been tested

Build container locally with precompiled crun v1.8.1 and tested in AWS EKS with and without an access to ECR

### CODE REVIEW

**Review by negz — COMMENTED** [review]
This looks good to me at first glance. Relates loosely to:

*  https://github.com/crossplane/crossplane/issues/3717

Could you please fill out the "how has this been tested" section of the PR template.

**Comment by jbw976** [comment]
Thanks a lot @AndrewChubatiuk for taking a stab at fixing these issues that have been blocking some early adopters progress on testing composition functions! That's very helpful 🤓 

> made it compatible with crun 1.8.1.

Do the changes to make it compatible with crun 1.8.1 have any impact to compatibility with older versions of crun? Should we be considering pinning the version of crun we use, as suggested by @negz [here](https://github.com/crossplane/crossplane/issues/3807#issuecomment-1450905774)?

**Comment by bobh66** [comment]
Great work @AndrewChubatiuk !  Thanks for digging into this!
Probably a good idea to pin crun to 1.8.1 as part of these changes, if that interface is going to be fragile.

File: internal/oci/pull.go [hunk 1]
```diff
@@ -141,11 +142,19 @@ type RemoteClient struct{}
...
 	opts := parse(o...)
+	if opts.pull == ImagePullPolicyNever {
+		return nil, errors.New(errPullNever)
+	}
 	if opts.auth != nil {
 		return remote.Image(ref, remote.WithContext(ctx), remote.WithAuth(opts.auth))
 	}
-	if opts.pull == ImagePullPolicyNever {
-		return nil, errors.New(errPullNever)
+	keychain, err := k8schain.NewInCluster(ctx, k8schain.Options{})
```
Comment (negz) [inline internal/oci/pull.go:151]: Will `NewInCluster` always fail if in-cluster auth isn't supported?

Or more generally, what causes `NewInCluster` to fail and what causes `remote.WithAuthFromKeychain` to fail? It may be worth adding comments to clarify.

File: internal/oci/pull.go [hunk 2]
```diff
@@ -141,11 +142,19 @@ type RemoteClient struct{}
...
+	var image ociv1.Image
+	if err == nil {
+		image, err = remote.Image(ref, remote.WithContext(ctx), remote.WithAuthFromKeychain(keychain))
+	}
+	if err == nil {
+		return image, nil
 	}
 	return remote.Image(ref, remote.WithContext(ctx))
```
Comment (negz) [inline internal/oci/pull.go:159]:
```suggestion
	keychain, err := k8schain.NewInCluster(ctx, k8schain.Options{})
	if err != nil {
		// In-cluster auth not supported. Fall back to no auth.
		return remote.Image(ref, remote.WithContext(ctx))
	}
	img, err := remote.Image(ref, remote.WithContext(ctx), remote.WithAuthFromKeychain(keychain))
	if err != nil {
		// In-cluster auth not supported. Fall back to no auth.	
		return remote.Image(ref, remote.WithContext(ctx))
	}
	return img, nil
```

Would this logic be equivalent? If so I'd prefer it, as I find it easier to follow.

**Review by negz — COMMENTED** [review]
Thanks @AndrewChubatiuk. A few comments, but this looks good overall.

Please squash/cleanup your git commits before we merge this (but please do keep the `sysfs` fix in a separate commit).

File: apis/apiextensions/fn/proto/v1alpha1/run_function.proto [hunk 3]
```diff
@@ -39,22 +39,11 @@ enum ImagePullPolicy {
...
-  string registry_token = 5;
-}
-
 // ImagePullConfig configures how a Composition Function container should be
 // pulled from a remote OCI registry.
 message ImagePullConfig {
   ImagePullPolicy pull_policy = 1;
-  ImagePullAuth auth = 2;
+  repeated string pull_secrets = 2;
 }
```
Comment (negz) [inline apis/apiextensions/fn/proto/v1alpha1/run_function.proto:47]: I don't think we can make this change. We need the Crossplane core pod, not `xfn` to resolve pull secrets into image pull auth credentials. This is because we expect other Composition Function runners will exist, and those runners won't always be running in the same Kubernetes cluster as the Crossplane pod (they might not be running in Kubernetes at all).

File: apis/apiextensions/v1/composition_types.go [hunk 4]
```diff
@@ -308,6 +312,14 @@ type ContainerFunction struct {
 	Runner *ContainerFunctionRunner `json:"runner,omitempty"`
 }
 
+// ContainerFunctionImagePullSecret specifies secrets which contain image credentials
+type ContainerFunctionImagePullSecret struct {
+	// Kubernetes secrets name for Composition Function images.
+	// Uses crossplane namespace and service account by default.
+	// +optional
+	Name string `json:"name,omitempty"`
```
Comment (negz) [inline apis/apiextensions/v1/composition_types.go:320]: Is this actually optional? It's the only field in the type. If it's not, please remove the `+optional` marker and the `omitempty` tag. (Same for the revision type.)

File: internal/xpkg/fetch.go [hunk 5]
```diff
@@ -135,7 +135,7 @@ func WithServiceAccount(sa string) FetcherOpt {
 func NewK8sFetcher(client kubernetes.Interface, opts ...FetcherOpt) (*K8sFetcher, error) {
 	k := &K8sFetcher{
 		client:    client,
-		transport: remote.DefaultTransport.Clone(),
+		transport: remote.DefaultTransport.(*http.Transport).Clone(),
```
Comment (negz) [inline internal/xpkg/fetch.go:138]: Why was this change needed?

**Review by negz — CHANGES_REQUESTED** [review]

File: apis/apiextensions/v1/composition_types.go [hunk 6]
```diff
@@ -295,6 +295,10 @@ type ContainerFunction struct {
 	// +kubebuilder:default="20s"
 	Timeout *metav1.Duration `json:"timeout,omitempty"`
 
+	// Secrets for pulling function images.
+	// +optional
+	ImagePullSecrets []ContainerFunctionImagePullSecret `json:"imagePullSecrets,omitempty"`
```
Comment (turkenh) [inline apis/apiextensions/v1/composition_types.go:300]:
```suggestion
	ImagePullSecrets []corev1.LocalObjectReference `json:"imagePullSecrets,omitempty"`
```

Any reason not to use the same type as [packagePullSecrets](https://github.com/crossplane/crossplane/blob/580d15ac8ac79e72acec7f95a70c8ddd9026d6b8/apis/pkg/v1/package_types.go#L43)?

File: internal/controller/apiextensions/composite/composition_ptf.go [hunk 7]
```diff
@@ -724,10 +730,34 @@ func RunFunction(ctx context.Context, fnio *iov1alpha1.FunctionIO, fn *v1.Contai
 		return nil, errors.Wrap(err, errDialRunner)
 	}
 
+	k8schainOpts := k8schain.Options{}
+	if sa := os.Getenv("POD_SERVICE_ACCOUNT"); sa != "" {
+		k8schainOpts.ServiceAccountName = sa
+	}
+	if ns := os.Getenv("POD_NAMESPACE"); ns != "" {
+		k8schainOpts.Namespace = ns
+	}
```
Comment (turkenh) [inline internal/controller/apiextensions/composite/composition_ptf.go:739]: We usually avoid reading environment values in arbitrary places in the codebase and, instead, parse them in all in a single location during startup and flow them down with options. 

The two variables here are already [parsed](https://github.com/crossplane/crossplane/blob/580d15ac8ac79e72acec7f95a70c8ddd9026d6b8/cmd/crossplane/core/core.go#L70-L71), and we should better pass them down [similar to how the package manager does](https://github.com/crossplane/crossplane/blob/580d15ac8ac79e72acec7f95a70c8ddd9026d6b8/cmd/crossplane/core/core.go#L171-L172).

File: internal/controller/apiextensions/composite/composition_ptf.go [hunk 8]
```diff
@@ -724,10 +730,34 @@ func RunFunction(ctx context.Context, fnio *iov1alpha1.FunctionIO, fn *v1.Contai
...
+	if fn.ImagePullSecrets != nil {
```
Comment (turkenh) [inline internal/controller/apiextensions/composite/composition_ptf.go:740]: We can drop this `if` check, the `for` loop below would skip if `nil`.

**Review by turkenh — COMMENTED** [review]
Thanks @AndrewChubatiuk, left a couple of comments!

I don't think this PR will be fixing https://github.com/crossplane/crossplane/issues/3718 though which is more about passing credentials for the function itself rather than pulling it image.

File: internal/controller/apiextensions/composite/composition_ptf.go [hunk 9]
```diff
@@ -724,10 +741,29 @@ func RunFunction(ctx context.Context, fnio *iov1alpha1.FunctionIO, fn *v1.Contai
...
+		k8schainOpts.ImagePullSecrets = append(k8schainOpts.ImagePullSecrets, ips.Name)
+	}
+	keychain, err := k8schain.NewInCluster(ctx, k8schainOpts)
+	if err != nil {
+		if errors.Is(err, rest.ErrNotInCluster) {
+			keychain = nil
+		} else {
+			return nil, err
+		}
+	}
```
Comment (negz) [inline internal/controller/apiextensions/composite/composition_ptf.go:758]:
```suggestion
	// If we're not in a cluster keychain will be nil.
	if err != nil && !errors.Is(err, rest.ErrNotInCluster) {
		return nil, errors.Wrap(err, errNewKeychain)
	}
```

I'd prefer this simpler implementation. I believe `keychain` should always be `nil` if it returns an error.

https://github.com/google/go-containerregistry/blob/v0.14.0/pkg/authn/k8schain/k8schain.go#L61

**Review by negz — CHANGES_REQUESTED** [review]
Sorry this review took so long! I had to switch focus for a while but I expect to be working on Composition Functions again for at least a few weeks. I do think this needs a few more changes - please ping me on Crossplane Slack when you're ready for me to take another look.

**Review by negz — APPROVED** [review]
Thanks for persisting with this @AndrewChubatiuk! It looks great. I have some ideas for small improvements, but I can open a follow-up PR for those. I'll copy you on that.
